"""add patient variant and metadata tables

Revision ID: 3226e581f26d
Revises: 
Create Date: 2026-02-25 13:56:13.676809

"""
from typing import Sequence, Union

from alembic import op
import sqlalchemy as sa


# revision identifiers, used by Alembic.
revision: str = '3226e581f26d'
down_revision: Union[str, Sequence[str], None] = None
branch_labels: Union[str, Sequence[str], None] = None
depends_on: Union[str, Sequence[str], None] = None


def upgrade() -> None:
    """Upgrade schema."""
    # ### commands auto generated by Alembic - please adjust! ###
    op.create_table('genes',
    sa.Column('id', sa.Integer(), autoincrement=True, nullable=False),
    sa.Column('symbol', sa.String(), nullable=False),
    sa.PrimaryKeyConstraint('id')
    )
    with op.batch_alter_table('genes', schema=None) as batch_op:
        batch_op.create_index(batch_op.f('ix_genes_symbol'), ['symbol'], unique=True)

    op.create_table('papers',
    sa.Column('id', sa.String(), nullable=False),
    sa.Column('gene_id', sa.Integer(), nullable=False),
    sa.Column('filename', sa.String(length=255), nullable=False),
    sa.Column('extraction_status', sa.Enum('PARSED', 'FAILED', 'QUEUED', name='extractionstatus'), server_default='QUEUED', nullable=False),
    sa.Column('pmid', sa.String(), nullable=True),
    sa.Column('pmcid', sa.String(), nullable=True),
    sa.Column('doi', sa.String(), nullable=True),
    sa.Column('title', sa.Text(), nullable=True),
    sa.Column('abstract', sa.Text(), nullable=True),
    sa.Column('journal', sa.String(), nullable=True),
    sa.Column('first_author', sa.String(), nullable=True),
    sa.Column('pub_year', sa.Integer(), nullable=True),
    sa.Column('citation', sa.Text(), nullable=True),
    sa.Column('is_open_access', sa.Boolean(), nullable=True),
    sa.Column('can_access', sa.Boolean(), nullable=True),
    sa.Column('license', sa.String(), nullable=True),
    sa.Column('link', sa.String(), nullable=True),
    sa.Column('paper_types', sa.JSON(), nullable=True),
    sa.Column('testing_methods', sa.JSON(), nullable=True),
    sa.Column('testing_methods_evidence', sa.JSON(), nullable=True),
    sa.ForeignKeyConstraint(['gene_id'], ['genes.id'], ondelete='CASCADE'),
    sa.PrimaryKeyConstraint('id')
    )
    with op.batch_alter_table('papers', schema=None) as batch_op:
        batch_op.create_index(batch_op.f('ix_papers_filename'), ['filename'], unique=False)
        batch_op.create_index(batch_op.f('ix_papers_gene_id'), ['gene_id'], unique=False)
        batch_op.create_index(batch_op.f('ix_papers_id'), ['id'], unique=False)

    op.create_table('patients',
    sa.Column('id', sa.Integer(), autoincrement=True, nullable=False),
    sa.Column('paper_id', sa.String(), nullable=False),
    sa.Column('identifier', sa.String(), nullable=False),
    sa.Column('proband_status', sa.String(), nullable=True),
    sa.Column('sex', sa.String(), nullable=True),
    sa.Column('age_diagnosis', sa.String(), nullable=True),
    sa.Column('age_report', sa.String(), nullable=True),
    sa.Column('age_death', sa.String(), nullable=True),
    sa.Column('country_of_origin', sa.String(), nullable=True),
    sa.Column('race_ethnicity', sa.String(), nullable=True),
    sa.Column('identifier_evidence', sa.Text(), nullable=True),
    sa.Column('sex_evidence', sa.Text(), nullable=True),
    sa.Column('age_diagnosis_evidence', sa.Text(), nullable=True),
    sa.Column('age_report_evidence', sa.Text(), nullable=True),
    sa.Column('age_death_evidence', sa.Text(), nullable=True),
    sa.Column('country_of_origin_evidence', sa.Text(), nullable=True),
    sa.Column('race_ethnicity_evidence', sa.Text(), nullable=True),
    sa.ForeignKeyConstraint(['paper_id'], ['papers.id'], ondelete='CASCADE'),
    sa.PrimaryKeyConstraint('id')
    )
    with op.batch_alter_table('patients', schema=None) as batch_op:
        batch_op.create_index(batch_op.f('ix_patients_paper_id'), ['paper_id'], unique=False)

    op.create_table('variants',
    sa.Column('id', sa.Integer(), autoincrement=True, nullable=False),
    sa.Column('paper_id', sa.String(), nullable=False),
    sa.Column('gene', sa.String(), nullable=False),
    sa.Column('transcript', sa.String(), nullable=True),
    sa.Column('protein_accession', sa.String(), nullable=True),
    sa.Column('genomic_accession', sa.String(), nullable=True),
    sa.Column('lrg_accession', sa.String(), nullable=True),
    sa.Column('gene_accession', sa.String(), nullable=True),
    sa.Column('variant_description_verbatim', sa.Text(), nullable=True),
    sa.Column('genomic_coordinates', sa.String(), nullable=True),
    sa.Column('genome_build', sa.String(), nullable=True),
    sa.Column('rsid', sa.String(), nullable=True),
    sa.Column('caid', sa.String(), nullable=True),
    sa.Column('hgvs_c', sa.String(), nullable=True),
    sa.Column('hgvs_p', sa.String(), nullable=True),
    sa.Column('hgvs_g', sa.String(), nullable=True),
    sa.Column('hgvs_c_inferred', sa.String(), nullable=True),
    sa.Column('hgvs_p_inferred', sa.String(), nullable=True),
    sa.Column('hgvs_p_inference_confidence', sa.String(), nullable=True),
    sa.Column('hgvs_p_inference_evidence_context', sa.Text(), nullable=True),
    sa.Column('hgvs_c_inference_confidence', sa.String(), nullable=True),
    sa.Column('hgvs_c_inference_evidence_context', sa.Text(), nullable=True),
    sa.Column('variant_type', sa.String(), nullable=True),
    sa.Column('variant_evidence_context', sa.Text(), nullable=True),
    sa.Column('variant_type_evidence_context', sa.Text(), nullable=True),
    sa.ForeignKeyConstraint(['paper_id'], ['papers.id'], ondelete='CASCADE'),
    sa.PrimaryKeyConstraint('id')
    )
    with op.batch_alter_table('variants', schema=None) as batch_op:
        batch_op.create_index(batch_op.f('ix_variants_paper_id'), ['paper_id'], unique=False)

    # ### end Alembic commands ###


def downgrade() -> None:
    """Downgrade schema."""
    # ### commands auto generated by Alembic - please adjust! ###
    with op.batch_alter_table('variants', schema=None) as batch_op:
        batch_op.drop_index(batch_op.f('ix_variants_paper_id'))

    op.drop_table('variants')
    with op.batch_alter_table('patients', schema=None) as batch_op:
        batch_op.drop_index(batch_op.f('ix_patients_paper_id'))

    op.drop_table('patients')
    with op.batch_alter_table('papers', schema=None) as batch_op:
        batch_op.drop_index(batch_op.f('ix_papers_id'))
        batch_op.drop_index(batch_op.f('ix_papers_gene_id'))
        batch_op.drop_index(batch_op.f('ix_papers_filename'))

    op.drop_table('papers')
    with op.batch_alter_table('genes', schema=None) as batch_op:
        batch_op.drop_index(batch_op.f('ix_genes_symbol'))

    op.drop_table('genes')
    # ### end Alembic commands ###
